File: apps/etf-service/src/services/market/service.test.ts
import assert from "node:assert/strict";
import test from "node:test";
import { SECURITIES } from "../../config/securities";
import type { DailyBar } from "../../routes/market/schema";
import {
  createMarketService,
  type MarketBarInput,
  type MarketCalendarInput,
  type MarketDataStore,
  type MarketSecurityRecord,
} from "./service";

const securityConfig = SECURITIES[0]!;

const securityRow: MarketSecurityRecord = {
  symbol: securityConfig.symbol,
  name: securityConfig.name,
  assetType: securityConfig.assetType,
  exchange: securityConfig.exchange,
  currency: securityConfig.currency,
  source: securityConfig.source,
  earliestTradeDate: "2013-12-31",
};

const createBar = (params: { tradeDate: string; close?: number }): DailyBar => ({
  symbol: securityConfig.symbol,
  adjType: "qfq",
  tradeDate: params.tradeDate,
  open: params.close ?? 100,
  high: params.close ?? 100,
  low: params.close ?? 100,
  close: params.close ?? 100,
  volume: 1,
  amount: 10,
  changeAmount: 0,
  changePercent: 0,
  rawWeekday: "星期五",
});

const createFakeStore = (params?: {
  bars?: DailyBar[];
  latestCachedTradeDate?: string | null;
  calendar?: Map<string, number>;
}) => {
  const calls = {
    upsertDailyBars: [] as Array<{ bars: MarketBarInput[]; exchange: string }>,
    upsertCalendarRows: [] as MarketCalendarInput[][],
    getCalendarMap: [] as Array<{ exchange: string; startDate: string; endDate: string }>,
  };
  const state = {
    bars: [...(params?.bars ?? [])],
    latestCachedTradeDate: params?.latestCachedTradeDate,
    calendar: params?.calendar ?? new Map<string, number>(),
  };

  const getLatestCachedTradeDate = (): string | null => {
    if (state.latestCachedTradeDate !== undefined) {
      return state.latestCachedTradeDate;
    }

    return (
      state.bars.map((bar) => bar.tradeDate).sort((left, right) => right.localeCompare(left))[0] ??
      null
    );
  };

  const store = {
    upsertSecurity: async () => {},
    upsertDailyBars: async (input) => {
      calls.upsertDailyBars.push(input);
      state.bars.push(
        ...input.bars.map((bar) => ({
          symbol: bar.symbol,
          adjType: bar.adjType ?? "none",
          tradeDate: bar.tradeDate,
          open: bar.open,
          high: bar.high,
          low: bar.low,
          close: bar.close,
          volume: bar.volume ?? 0,
          amount: bar.amount ?? 0,
          changeAmount: bar.changeAmount ?? 0,
          changePercent: bar.changePercent ?? 0,
          rawWeekday: bar.rawWeekday ?? "",
        })),
      );
      state.latestCachedTradeDate = undefined;
    },
    upsertCalendarRows: async (rows) => {
      calls.upsertCalendarRows.push(rows);
      for (const row of rows) {
        state.calendar.set(row.tradeDate, row.isOpen);
      }
    },
    listSecuritiesRows: async () => [securityRow],
    getSecurityRow: async (symbol) => (symbol === securityRow.symbol ? securityRow : null),
    getLatestCachedTradeDate: async () => getLatestCachedTradeDate(),
    listBars: async (input) =>
      state.bars
        .filter(
          (bar) =>
            bar.symbol === input.symbol &&
            bar.adjType === input.adjType &&
            bar.tradeDate >= input.startDate &&
            bar.tradeDate <= input.endDate,
        )
        .sort((left, right) => left.tradeDate.localeCompare(right.tradeDate)),
    getCalendarMap: async (input) => {
      calls.getCalendarMap.push(input);
      return state.calendar;
    },
  } satisfies MarketDataStore;

  return { store, calls };
};

void test("日期完全早于最早数据时不访问远端且返回无效范围", async () => {
  const { store, calls } = createFakeStore({ latestCachedTradeDate: null });
  let fetchCount = 0;
  const service = createMarketService({
    securities: SECURITIES,
    store,
    fetchRemoteKlineBars: async () => {
      fetchCount += 1;
      return { bars: [] };
    },
    getTMinusOneDate: () => "2026-05-30",
  });

  const response = await service.getDailyBars({
    symbol: securityConfig.symbol,
    adjType: "qfq",
    startDate: "2012-01-01",
    endDate: "2013-01-01",
  });

  assert.equal(response.meta.cacheStatus, "invalid");
  assert.equal(response.meta.refreshed, false);
  assert.equal(response.bars.length, 0);
  assert.equal(fetchCount, 0);
  assert.equal(calls.getCalendarMap.length, 0);
  assert.equal(calls.upsertDailyBars.length, 0);
});

void test("数据库缓存覆盖请求区间时直接返回缓存数据", async () => {
  const { store, calls } = createFakeStore({
    bars: [createBar({ tradeDate: "2026-05-29", close: 101 })],
    latestCachedTradeDate: "2026-05-29",
  });
  let fetchCount = 0;
  const service = createMarketService({
    securities: SECURITIES,
    store,
    fetchRemoteKlineBars: async () => {
      fetchCount += 1;
      return { bars: [] };
    },
    getTMinusOneDate: () => "2026-05-30",
  });

  const response = await service.getDailyBars({
    symbol: securityConfig.symbol,
    adjType: "qfq",
    startDate: "2026-05-29",
    endDate: "2026-05-30",
  });

  assert.equal(response.meta.cacheStatus, "cache");
  assert.equal(response.meta.refreshed, false);
  assert.equal(response.bars.length, 1);
  assert.equal(response.bars[0]?.tradeDate, "2026-05-29");
  assert.equal(fetchCount, 0);
  assert.equal(calls.upsertDailyBars.length, 0);
});

void test("数据库缺少有效交易日时拉取远端并过滤当天之后的数据", async () => {
  const { store, calls } = createFakeStore({
    bars: [createBar({ tradeDate: "2026-05-28", close: 100 })],
    latestCachedTradeDate: "2026-05-28",
  });
  const remoteBars: MarketBarInput[] = [
    createBar({ tradeDate: "2026-05-29", close: 101 }),
    createBar({ tradeDate: "2026-05-31", close: 102 }),
  ];
  let fetchCount = 0;
  const service = createMarketService({
    securities: SECURITIES,
    store,
    fetchRemoteKlineBars: async () => {
      fetchCount += 1;
      return { bars: remoteBars };
    },
    getTMinusOneDate: () => "2026-05-30",
  });

  const response = await service.getDailyBars({
    symbol: securityConfig.symbol,
    adjType: "qfq",
    startDate: "2026-05-28",
    endDate: "2026-05-31",
  });

  assert.equal(response.meta.cacheStatus, "refreshed");
  assert.equal(response.meta.refreshed, true);
  assert.equal(response.meta.effectiveEndDate, "2026-05-30");
  assert.equal(fetchCount, 1);
  assert.equal(calls.upsertDailyBars.length, 1);
  assert.deepEqual(
    calls.upsertDailyBars[0]?.bars.map((bar) => bar.tradeDate),
    ["2026-05-29"],
  );
  assert.deepEqual(
    response.bars.map((bar) => bar.tradeDate),
    ["2026-05-28", "2026-05-29"],
  );
});

void test("证券列表会返回每只证券的最早日期和最新缓存日期", async () => {
  const { store } = createFakeStore({
    bars: [createBar({ tradeDate: "2026-05-29", close: 101 })],
  });
  const service = createMarketService({
    securities: SECURITIES,
    store,
    fetchRemoteKlineBars: async () => ({ bars: [] }),
    getTMinusOneDate: () => "2026-05-30",
  });

  const securities = await service.listSecurities();

  assert.equal(securities.length, 1);
  assert.equal(securities[0]?.symbol, securityConfig.symbol);
  assert.equal(securities[0]?.earliestTradeDate, "2013-12-31");
  assert.equal(securities[0]?.latestCachedTradeDate, "2026-05-29");
});

File: apps/etf-service/internal/modules/market/service_test.go
package market

import (
	"context"
	"testing"
	"time"

	"kxh-awesome/etf-service/internal/shared/config"
)

var testSecurity = config.Securities[0]

var testSecurityRow = SecurityRecord{
	Symbol:            testSecurity.Symbol,
	Name:              testSecurity.Name,
	AssetType:         testSecurity.AssetType,
	Exchange:          &testSecurity.Exchange,
	Currency:          testSecurity.Currency,
	Source:            &testSecurity.Source,
	EarliestTradeDate: "2013-12-31",
}

type fakeStore struct {
	bars                  []DailyBar
	latestCachedTradeDate *string
	calendar              map[string]int
	upsertDailyBarsCalls  [][]MarketBarInput
	upsertCalendarCalls   [][]CalendarInput
	getCalendarMapCalls   int
}

func (s *fakeStore) UpsertDailyBars(_ context.Context, bars []MarketBarInput, _ string) error {
	s.upsertDailyBarsCalls = append(s.upsertDailyBarsCalls, bars)
	for _, bar := range bars {
		s.bars = append(s.bars, DailyBar(bar))
	}
	s.latestCachedTradeDate = nil
	return nil
}

func (s *fakeStore) UpsertCalendarRows(_ context.Context, rows []CalendarInput) error {
	s.upsertCalendarCalls = append(s.upsertCalendarCalls, rows)
	if s.calendar == nil {
		s.calendar = map[string]int{}
	}
	for _, row := range rows {
		s.calendar[row.TradeDate] = row.IsOpen
	}
	return nil
}

func (s *fakeStore) ListSecuritiesRows(_ context.Context) ([]SecurityRecord, error) {
	return []SecurityRecord{testSecurityRow}, nil
}

func (s *fakeStore) GetSecurityRow(_ context.Context, symbol string) (*SecurityRecord, error) {
	if symbol != testSecurity.Symbol {
		return nil, nil
	}
	row := testSecurityRow
	return &row, nil
}

func (s *fakeStore) GetLatestCachedTradeDate(_ context.Context, _ string, _ string) (*string, error) {
	if s.latestCachedTradeDate != nil {
		return s.latestCachedTradeDate, nil
	}
	var latest *string
	for _, bar := range s.bars {
		if latest == nil || bar.TradeDate > *latest {
			tradeDate := bar.TradeDate
			latest = &tradeDate
		}
	}
	return latest, nil
}

func (s *fakeStore) ListBars(_ context.Context, symbol string, adjType string, startDate string, endDate string) ([]DailyBar, error) {
	var result []DailyBar
	for _, bar := range s.bars {
		if bar.Symbol == symbol && bar.AdjType == adjType && bar.TradeDate >= startDate && bar.TradeDate <= endDate {
			result = append(result, bar)
		}
	}
	return result, nil
}

func (s *fakeStore) GetCalendarMap(_ context.Context, _ string, _ string, _ string) (map[string]int, error) {
	s.getCalendarMapCalls++
	if s.calendar == nil {
		return map[string]int{}, nil
	}
	return s.calendar, nil
}

type fakeFetcher struct {
	bars  []MarketBarInput
	calls int
}

func (f *fakeFetcher) FetchRemoteKlineBars(_ context.Context, _ config.SecurityConfig) ([]MarketBarInput, error) {
	f.calls++
	return f.bars, nil
}

func TestGetDailyBarsInvalidRangeDoesNotFetch(t *testing.T) {
	store := &fakeStore{}
	fetcher := &fakeFetcher{}
	service := NewMarketServiceWithNow(config.Securities, store, fetcher, fixedNow)

	response, err := service.GetDailyBars(context.Background(), GetDailyBarsRequest{
		Symbol:    testSecurity.Symbol,
		AdjType:   "qfq",
		StartDate: strPtr("2012-01-01"),
		EndDate:   strPtr("2013-01-01"),
	})
	if err != nil {
		t.Fatalf("GetDailyBars returned error: %v", err)
	}
	if response.Meta.CacheStatus != "invalid" {
		t.Fatalf("CacheStatus = %s, want invalid", response.Meta.CacheStatus)
	}
	if response.Meta.Refreshed {
		t.Fatal("Refreshed = true, want false")
	}
	if len(response.Bars) != 0 {
		t.Fatalf("len(Bars) = %d, want 0", len(response.Bars))
	}
	if fetcher.calls != 0 {
		t.Fatalf("fetch calls = %d, want 0", fetcher.calls)
	}
	if store.getCalendarMapCalls != 0 {
		t.Fatalf("calendar calls = %d, want 0", store.getCalendarMapCalls)
	}
}

func TestGetDailyBarsReturnsCacheWhenCovered(t *testing.T) {
	latest := "2026-05-29"
	store := &fakeStore{
		bars:                  []DailyBar{createBar("2026-05-29", 101)},
		latestCachedTradeDate: &latest,
	}
	fetcher := &fakeFetcher{}
	service := NewMarketServiceWithNow(config.Securities, store, fetcher, fixedNow)

	response, err := service.GetDailyBars(context.Background(), GetDailyBarsRequest{
		Symbol:    testSecurity.Symbol,
		AdjType:   "qfq",
		StartDate: strPtr("2026-05-29"),
		EndDate:   strPtr("2026-05-30"),
	})
	if err != nil {
		t.Fatalf("GetDailyBars returned error: %v", err)
	}
	if response.Meta.CacheStatus != "cache" {
		t.Fatalf("CacheStatus = %s, want cache", response.Meta.CacheStatus)
	}
	if fetcher.calls != 0 {
		t.Fatalf("fetch calls = %d, want 0", fetcher.calls)
	}
	if len(response.Bars) != 1 || response.Bars[0].TradeDate != "2026-05-29" {
		t.Fatalf("bars = %#v, want one cached bar", response.Bars)
	}
}

func TestGetDailyBarsRefreshesAndFiltersAfterTMinusOne(t *testing.T) {
	latest := "2026-05-28"
	store := &fakeStore{
		bars:                  []DailyBar{createBar("2026-05-28", 100)},
		latestCachedTradeDate: &latest,
	}
	fetcher := &fakeFetcher{
		bars: []MarketBarInput{
			createBar("2026-05-29", 101),
			createBar("2026-05-31", 102),
		},
	}
	service := NewMarketServiceWithNow(config.Securities, store, fetcher, fixedNow)

	response, err := service.GetDailyBars(context.Background(), GetDailyBarsRequest{
		Symbol:    testSecurity.Symbol,
		AdjType:   "qfq",
		StartDate: strPtr("2026-05-28"),
		EndDate:   strPtr("2026-05-31"),
	})
	if err != nil {
		t.Fatalf("GetDailyBars returned error: %v", err)
	}
	if response.Meta.CacheStatus != "refreshed" {
		t.Fatalf("CacheStatus = %s, want refreshed", response.Meta.CacheStatus)
	}
	if response.Meta.EffectiveEndDate == nil || *response.Meta.EffectiveEndDate != "2026-05-30" {
		t.Fatalf("EffectiveEndDate = %v, want 2026-05-30", response.Meta.EffectiveEndDate)
	}
	if fetcher.calls != 1 {
		t.Fatalf("fetch calls = %d, want 1", fetcher.calls)
	}
	if len(store.upsertDailyBarsCalls) != 1 {
		t.Fatalf("upsert calls = %d, want 1", len(store.upsertDailyBarsCalls))
	}
	if got := store.upsertDailyBarsCalls[0]; len(got) != 1 || got[0].TradeDate != "2026-05-29" {
		t.Fatalf("upserted bars = %#v, want only 2026-05-29", got)
	}
	if len(response.Bars) != 2 || response.Bars[0].TradeDate != "2026-05-28" || response.Bars[1].TradeDate != "2026-05-29" {
		t.Fatalf("response bars = %#v, want cached plus refreshed bar", response.Bars)
	}
}

func TestListSecuritiesIncludesLatestCachedTradeDate(t *testing.T) {
	store := &fakeStore{bars: []DailyBar{createBar("2026-05-29", 101)}}
	service := NewMarketServiceWithNow(config.Securities, store, &fakeFetcher{}, fixedNow)

	securities, err := service.ListSecurities(context.Background())
	if err != nil {
		t.Fatalf("ListSecurities returned error: %v", err)
	}
	if len(securities) != 1 {
		t.Fatalf("len(securities) = %d, want 1", len(securities))
	}
	if securities[0].Symbol != testSecurity.Symbol {
		t.Fatalf("Symbol = %s, want %s", securities[0].Symbol, testSecurity.Symbol)
	}
	if securities[0].EarliestTradeDate != "2013-12-31" {
		t.Fatalf("EarliestTradeDate = %s, want 2013-12-31", securities[0].EarliestTradeDate)
	}
	if securities[0].LatestCachedTradeDate == nil || *securities[0].LatestCachedTradeDate != "2026-05-29" {
		t.Fatalf("LatestCachedTradeDate = %v, want 2026-05-29", securities[0].LatestCachedTradeDate)
	}
}

func fixedNow() time.Time {
	return time.Date(2026, 5, 31, 12, 0, 0, 0, time.UTC)
}

func createBar(tradeDate string, closeValue float64) DailyBar {
	return DailyBar{
		Symbol:        testSecurity.Symbol,
		AdjType:       "qfq",
		TradeDate:     tradeDate,
		Open:          closeValue,
		High:          closeValue,
		Low:           closeValue,
		Close:         closeValue,
		Volume:        1,
		Amount:        10,
		ChangeAmount:  0,
		ChangePercent: 0,
		RawWeekday:    "星期五",
	}
}

func strPtr(value string) *string {
	return &value
}
